src/nsrt_learning.py

```diff
@@ -248,8 +244,102 @@ def _learn_pnad_side_predicates(
         ground_atom_dataset: List[GroundAtomTrajectory],
         train_tasks: List[Task], predicates: Set[Predicate],
         segments: List[Segment],
-        verbose: bool) -> List[PartialNSRTAndDatastore]:
-    raise NotImplementedError
+        segmented_trajs: List[List[Segment]]) -> List[PartialNSRTAndDatastore]:
+    # There are no goal states for this search; run until exhausted.
+    def _check_goal(s: Tuple[PartialNSRTAndDatastore, ...]) -> bool:
+        del s  # unused
+        return False
+
+    # Consider changing each effect to a side predicate.
+    def _get_successors(
+        s: Tuple[PartialNSRTAndDatastore, ...],
+    ) -> Iterator[Tuple[None, Tuple[PartialNSRTAndDatastore, ...], float]]:
+        for i in range(len(s)):
+            pnad = s[i]
+            _, option_vars = pnad.option_spec
+            for effect_set, add_or_delete in [(pnad.op.add_effects, "add"),
+                                              (pnad.op.delete_effects,
+                                               "delete")]:
+                for effect in effect_set:
+                    new_pnad = PartialNSRTAndDatastore(
+                        pnad.op.effect_to_side_predicate(
+                            effect, option_vars, add_or_delete),
+                        pnad.datastore, pnad.option_spec)
+                    sprime = list(s)
+                    sprime[i] = new_pnad
+                    yield (None, tuple(sprime), 1.0)
+
+    # Hard to avoid this circular import, since approaches need nsrt_learning.
+    from predicators.src.approaches.grammar_search_invention_approach import \
```

Comment:
let's try to resolve this before merging

Reply:
i don't have the slightest idea how. any suggestions? even if we move this side predicate stuff into a new file it won't help, bc approaches -> nsrt_learning -> side_predicate_learning will be a cycle

Reply:
resolved in #454